Carousel detail page

AmauryMi8 opened the pull request:
Carousel feature for Star System details display

Commit:
Carousel view

Init preview feature for Droid and IOS

Commit:
carousel

init

Commit:
fqdgqefy

qdhqdgh

Commit:
carousel : retreive data

retreive data from api, etc. and copy them 3 times (one time per tab) to display details info on separate tab.
Remove indicator for now because of an error.
try to update xamarin

Commit:
carousel

add indicator to carousel

Commit:
add dataSelectorTemplate

Commit:
try to implement datatemplateSelector

still have an error...

Commit:
Carousel : now working like a charm

Commit:
carousel : finished

Commit:
detail view : some adjust...

stiiifff commented on challengeMk2/DataSelectors/SystemDetailTemplateSelector.cs:
```diff
@@ -0,0 +1,40 @@
+                    return PrimaryStarInfoTemplate;
+                default:
+                    return GeneralInfoTemplate;
+            }
```
It's possible to simplify this code by defining a Dictionary with all the possibilities, and just doing a lookup into it using the data selector ID.

stiiifff commented on challengeMk2/DataSelectors/SystemDetailTemplateSelector.cs:
```diff
@@ -0,0 +1,40 @@
+            //    1 => SystemInfoTemplate,
+            //    2 => PrimaryStarInfoTemplate,
+            //    _ => GeneralInfoTemplate,
+            //};
```
Please remove unnecessary comments.

stiiifff commented on challengeMk2/Models/StarSystem.cs:
```diff
@@ -29,6 +29,16 @@ public class StarSystem
+        //{
+        //    get => dataSelectorID;
+        //    set => dataSelectorID = value;
+        //}
```
Please remove unnecessary comments.

stiiifff commented on challengeMk2/ViewModels/SystemDetailCarouselViewModel.cs:
```diff
@@ -0,0 +1,164 @@
+using Xamarin.Forms;
+
+
+namespace challengeMk2.ViewModels
```
Namespaces should always use Pascal Case.

stiiifff commented on challengeMk2/ViewModels/SystemDetailCarouselViewModel.cs:
```diff
@@ -0,0 +1,164 @@
+
+
+
+
```
Cleanup unnecessary spaces.

stiiifff commented on challengeMk2/ViewModels/SystemDetailCarouselViewModel.cs:
```diff
@@ -0,0 +1,164 @@
+
+
+
+
```
Cleanup unnecessary spaces.

stiiifff commented on challengeMk2/ViewModels/SystemDetailCarouselViewModel.cs:
```diff
@@ -0,0 +1,164 @@
+
+        public NetworkAccess CurrentConnectivity { get; set; }
+
+        public IList SystemInfos { get; set; }
```
It's OK to declare it as ObservableCollection instead of a generic IList.

stiiifff commented on challengeMk2/ViewModels/SystemDetailCarouselViewModel.cs:
```diff
@@ -0,0 +1,164 @@
+
+        private async Task<StarSystem> GetDetailsFromApi()
+        {
+            HttpClient client = new HttpClient();
```
Don't forget to dispose the HttpClient. You could use the C# 8.0 using declaration syntax.

AmauryMi8 replied:
I came across this article about HttpClient : https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/
What do you thing is better, use using or a static HttpClient ?

stiiifff commented on challengeMk2/ViewModels/SystemDetailCarouselViewModel.cs:
```diff
@@ -0,0 +1,164 @@
+            systemWithId.PermitName = detailedSystem.PermitName;
+            systemWithId.Information = detailedSystem.Information;
+            systemWithId.PrimaryStar = detailedSystem.PrimaryStar;
+            systemWithId.Coords = detailedSystem.Coords;
```
Use inline constructor property assignment.

stiiifff commented on challengeMk2/ViewModels/SystemDetailCarouselViewModel.cs:
```diff
@@ -0,0 +1,164 @@
+            currentSystem = selectedSystem;
+
+            //Get infos
+            UpdateSystemData();
```
It's not a good pattern to initialize data in the constructor.
More generally, a constructor should never perform any expensive work (e.g. I/O), and is generally not expected to fail (except for basic argument validation).

stiiifff commented on challengeMk2/ViewModels/SystemDetailCarouselViewModel.cs:
```diff
@@ -0,0 +1,164 @@
+
+
+
+        private async void UpdateSystemData()
```
Make sure you know what you're doing when using `async void` ...

stiiifff commented on challengeMk2/Views/StarSystemsPage.xaml.cs:
```diff
@@ -21,7 +21,8 @@ protected override void OnBindingContextChanged()
             if (vm!= null)
             {
-                vm.NavigateTodetailPage = async (starSystem) => await Navigation.PushAsync(new SystemDetailPage(new SystemDetailViewModel(starSystem)));
+                //vm.NavigateTodetailPage = async (starSystem) => await Navigation.PushAsync(new SystemDetailPage(new SystemDetailViewModel(starSystem)));
```
Cleanup old commented code.

stiiifff commented on challengeMk2/Views/SystemDetailCarouselPage.xaml:
```diff
@@ -0,0 +1,188 @@
+        <ResourceDictionary>
+
+            <!--GENERAL INFO-->
+            <DataTemplate x:Key="GeneralTemplate">
```
Would be nice to extract the inline templates into separate XAML components and refer to them here.

Commit:
Template selector

Simplify method with a dictionary <int, datatemplate> and remove switch()

Commit:
StarSystem class

Clean up code

Commit:
Pascal case for Namespace

Commit:
fix error (pascal case)

Commit:
Pascal case for NameSpace retry (the good one)

Commit:
Update SystemDetailCarouselViewModel.cs

Cleanup unnecessary spaces

Commit:
Update SystemDetailCarouselViewModel.cs

Declare SystemInfos as ObservableCollection instead of IList

Commit:
Update SystemDetailCarouselViewModel.cs

Dispose HttpClient with using declaration

Commit:
Update SystemDetailCarouselViewModel.cs

Use inline constructor property assignment

Commit:
System detail carousel data init

remove data initialization from Ctor and put it in the code behind

Commit:
Update StarSystemsPage.xaml.cs

Clean up old commented code

CtrlZzzz committed:
Fix build error

fix Upper case

CtrlZzzz committed:
Carousel xaml DataTemplates

DataTemplate are now in 3 separate xaml files and be refered in carousel page for dataSelector.

CtrlZzzz committed:
Update SystemDetailCarouselPage.xaml

REmove Converter sinds it is in the app.xaml

CtrlZzzz committed:
Start working on Theme

add themes folder
add defaultTheme file

CtrlZzzz committed:
Update DefaultTheme.xaml

Add Colors

CtrlZzzz committed:
Add resources

Banners and Bgs

Commit:
Clean up

Commit:
Update DefaultTheme.xaml

add colors

Commit:
Update DefaultTheme.xaml

Commit:
Update App.xaml

Move Converter out of the resource Dictionary in the application resources root. allow to use merged dictionary only for the theme.

Commit:
Add Some Theme

Add DarkTheme Test

Commit:
Try to resolve resource not find...

Commit:
fix build errors

Rename some properties in .sln and .csproj files with the GOOD name space...

Commit:
Update Theme

Add Default theme (Dark theme)
Customize colors

Commit:
Customize Status bar

Change color from status bar on Android and IOS

Commit:
Change default background color android

Commit:
Theme customization

Change background color from collection items if selected or not

Commit:
Update StarSystemsPage.xaml

Tweek Theme

Commit:
Custom Control

Starting to work on a custom control with bindableProperties.

Commit:
CleanUp();

Commit:
Working on Custom control

Commit:
finish control

Commit:
Clean up DualLabel

But , atm, stay with OnPropertychanged because  there no more binding if removed. Checking why.

Commit:
Layout system details pages

Commit:
some

Commit:
Update DualLabel.cs

Try to change text alignment in function of horizontalOptions

Commit:
adjustments on layout

Commit:
Add banner switch when carouselview position changes

Commit:
restrict device orientation to portrait

Commit:
Clean up DualLabel controller

Removed all OnPropertyChanged from the controller and add binding instead

Commit:
upgrade project to c# 8

=> change "Using" implementations with c# 8 version

Commit:
Update DualLabel.cs

Adapt text horizontal alignement to its horizontalOptions
